clang/lib/StaticAnalyzer/Checkers/BuiltinFunctionChecker.cpp

```diff
@@ -50,6 +118,75 @@ class BuiltinFunctionChecker : public Checker<eval::Call> {
 
 } // namespace
 
+std::pair<bool, bool>
+BuiltinFunctionChecker::checkOverflow(CheckerContext &C, SVal RetVal,
+                                      QualType Res) const {
+  ProgramStateRef State = C.getState();
+  SValBuilder &SVB = C.getSValBuilder();
+  ASTContext &ACtx = C.getASTContext();
+
+  // Calling a builtin with a non-integer type result produces compiler error.
+  assert(Res->isIntegerType());
+
+  unsigned BitWidth = ACtx.getIntWidth(Res);
+  auto MinVal =
+      llvm::APSInt::getMinValue(BitWidth, Res->isUnsignedIntegerType());
+  auto MaxVal =
+      llvm::APSInt::getMaxValue(BitWidth, Res->isUnsignedIntegerType());
+
+  SVal IsLeMax =
+      SVB.evalBinOp(State, BO_LE, RetVal, nonloc::ConcreteInt(MaxVal), Res);
+  SVal IsGeMin =
+      SVB.evalBinOp(State, BO_GE, RetVal, nonloc::ConcreteInt(MinVal), Res);
+
+  auto [MayNotOverflow, MayOverflow] =
+      State->assume(IsLeMax.castAs<DefinedOrUnknownSVal>());
+  auto [MayNotUnderflow, MayUnderflow] =
+      State->assume(IsGeMin.castAs<DefinedOrUnknownSVal>());
+
+  return {MayOverflow || MayUnderflow, MayNotOverflow && MayNotUnderflow};
+}
+
+void BuiltinFunctionChecker::handleOverflowBuiltin(const CallEvent &Call,
+                                                   CheckerContext &C,
+                                                   BinaryOperator::Opcode Op,
+                                                   QualType ResultType) const {
+  // Calling a builtin with an incorrect argument count produces compiler error.
+  assert(Call.getNumArgs() == 3);
+
+  ProgramStateRef State = C.getState();
+  SValBuilder &SVB = C.getSValBuilder();
+  const Expr *CE = Call.getOriginExpr();
+
+  SVal Arg1 = Call.getArgSVal(0);
+  SVal Arg2 = Call.getArgSVal(1);
+
+  SVal RetValMax = SVB.evalBinOp(State, Op, Arg1, Arg2,
+                                 getSufficientTypeForOverflowOp(C, ResultType));
+  SVal RetVal = SVB.evalBinOp(State, Op, Arg1, Arg2, ResultType);
+
+  auto [Overflow, NotOverflow] = checkOverflow(C, RetValMax, ResultType);
+  if (NotOverflow) {
+    ProgramStateRef StateNoOverflow =
+        State->BindExpr(CE, C.getLocationContext(), SVB.makeTruthVal(false));
+
+    if (auto L = Call.getArgSVal(2).getAs<Loc>()) {
+      StateNoOverflow =
+          StateNoOverflow->bindLoc(*L, RetVal, C.getLocationContext());
+
+      // Propagate taint if any of the argumets were tainted
+      if (isTainted(State, Arg1) || isTainted(State, Arg2))
+        StateNoOverflow = addTaint(StateNoOverflow, *L);
+    }
+
+    C.addTransition(StateNoOverflow);
```

Comment:
This transition should add a NoteTag highlighting that we assumed to have no overflow iff the StateOverflow was also valid. In that case, it should probably also call trackExpressionValue on both or some of the operands that played a role in that decision.
@NagyDonat is usually against trackExprVal, so I'm tagging him.

Reply:
Thanks for the ping, but I think that in this concrete case `trackExpressionValue` is probably OK.

I mostly advocated against its use when we were talking about ambitious plans that would've required extending the current code and/or using it in new, unusual situations. The code of `trackExpressionValue` is ugly and convoluted, but it is working (in the sense of "it's working, don't touch it").


By the way, if you want to explain the origin of _symbolic_ values, then calling `markInteresting()` on the relevant symbols is another option that may be sufficient for your goals. It is less general than `trackExpressionValue()` (e.g. it doesn't track concrete values, it doesn't track control dependencies etc.) but it doesn't apply random heuristics which are not relevant for you.

Reply:
Yes. To me, here the value is the important thing to track, but that said, we are likely more interested in tracking the taint flow, so markInteresting should be indeed enough. Thanks.

Reply:
Sorry, I am not sure I get it correctly (I am still learning internals, so might be dumb question). 

As far as I see, `markInteresting` is called on bug reported object. And there is no bug report object, since there is no bug here, we just do assumtions. I tried to find an example in other checkers, but I don't see it.

(Like I see manipulations with `markInteresting` only during construction of bug report).

Reply:
Thank you for explanation!

Ok, so based on my local testing, if overflow happens, then result does not end up in `InterestingSymbols` in bug reporter, so there is no need to note that condition. I tried different tests, but anyway it does not appear in this list.


I ended up just covering non-overflow case in callback.

Reply:
> if overflow happens, then result does not end up in InterestingSymbols in bug reporter,

Oops, I wanted to refer to the _result of the operation_ (which is written to the memory area specified by the third argument), not the boolean "did overflow happen?" return value of the `__builitin_X_overflow()`. However, now that I think about this, we should display the note if either of these values are interesting.

Reply:
> Oops, I wanted to refer to the result of the operation (which is written to the memory area specified by the third argument), not the boolean "did overflow happen?"

Yes, that was clear and I implemented it in the lasted version, but anyway in case of say

```c
int res;
if (__builtin_add_oveflow(a, b, &res))
   doStmh(res); // <-- read of uninit value
``` 

`res` is somehow not interesting... =(

Reply:
Oh I see... the undefined rvalue within `res` is represented by an `UndefinedVal`, which is not symbolic (does not have a unique identity, unlike e.g. `nonloc::SymbolVal`), so we cannot mark it as interesting, because it does not have an associated symbol.

Either always display the "Assuming overflow happens" note (it is a surprising case that's worth some highlight and with `IsPrunable` set to true it won't be _too_ spammy), or leave this as-is for now and perhaps return with a better solution in a follow-up  commit.

Reply:
Added notes for overflow case with prunable flag in the latest version, thanks!